### ISSUE TITLE
[torchx/k8s] support device mounts as device plugin + resource limits

### ISSUE THREAD

**Issue by kiukchung** [pytorch/torchx#443]
## Description
Instead of supporting device mounts as volume mounts under privileged mode, support it as a device plugin that needs to be installed on the k8s cluster and add a way to specify resource/limits.

## Motivation/Background
On k8s the docker equivalent of `--device` is supported as [k8s device plugin](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/). This is to support mounting the host's `/dev/**` on the container (just regular bind/volume mount won't work unless under privileged mode which not so great from a security standpoint).

TorchX exposes these device mounts as `Role.mounts` of type `DeviceMount` (see `torchx/specs/api.py`) and for consistency sake supports device mounts on k8s via privileged volume mounts (but as mentioned before has security implications).

## Detailed Proposal

For kubernetes scheduler change the current device mount logic from privileged volume mount to:

1. (prerequisite) have the user install the k8s device plugin for the accelerator they want to device mount
2.  expose `containers/resource/limit` as the [Role.metadata](https://github.com/pytorch/torchx/blob/main/torchx/specs/api.py#L294) under some key (e.g. "k8s.resource.limits")

Basically the resource limit that would be specified in the specs.yaml:

```
---
apiVersion: v1
kind: Pod
metadata:
  name: demo-pod
spec:
  containers:
    - name: demo-container-1
      image: k8s.gcr.io/pause:2.0
      resources:
        limits:
          hardware-vendor.example/foo: 2
```

For the device plugin for the device `example/foo` from `hardware-vendor` would be specified in the `Role` as

```
Role.metadata = {
    "k8s.container.resource.limits:hardware-vendor.example/foo" : 2,
}
```

NOTE: the "key" does not necessarily have to be `k8s.container.resource.limits:hardware-vendor.example/foo`, the implementor can figure out a better scheme.

## Alternatives

The resource/limit can be read off of any of:

1. `Role.metadata` (shown above)
2. `Role.resource.capabilities`
3. `DeviceMount.kubernetes_resource_limits` : this would be an added parameter to DeviceMount which can be configured optionally to support the same device mount on multiple schedulers (and keeps everything tidy in one place).  

Example usage of option 3 would be:

```
Role.mounts = [
  DeviceMount(
    # src and dst paths as usual
    src_path="/dev/infiniband/uverbs0", 
    dst_path="/dev/infiniband/uverbs0", 
    # same spec but in k8s lingo, if specified any Role with this device mount would also support running on k8s
    kubernetes_resource_limits=("vpc.amazonaws.com/efa", 1) # type Tuple[str, int]
  )
]
``` 

If the host had 4 EFA adapters:
```
Role.mounts = [
  DeviceMount(
      src_path="/dev/infiniband/uverbs0", 
    dst_path="/dev/infiniband/uverbs0", 
    kubernetes_resource_limits=("vpc.amazonaws.com/efa", 1) # type Tuple[str, int]
  ),
  DeviceMount(
    src_path="/dev/infiniband/uverbs1", 
    dst_path="/dev/infiniband/uverbs1", 
    kubernetes_resource_limits=("vpc.amazonaws.com/efa", 1) # type Tuple[str, int]
  ),
  DeviceMount(
    src_path="/dev/infiniband/uverbs2", 
    dst_path="/dev/infiniband/uverbs2", 
    kubernetes_resource_limits=("vpc.amazonaws.com/efa", 1) # type Tuple[str, int]
  ),
  DeviceMount(
    src_path="/dev/infiniband/uverbs3", 
    dst_path="/dev/infiniband/uverbs3", 
    kubernetes_resource_limits=("vpc.amazonaws.com/efa", 1) # type Tuple[str, int]
  ),
]
```

and in the torchx.schedulers.kubernetes_scheduler we'd have to add logic that groups by the unique keys in the `kubernetes_resource_limits` across all device mounts and add up the values so that ultimately we can pass it as

```
---
apiVersion: v1
kind: Pod
metadata:
  name: demo-pod
spec:
  containers:
    - name: demo-container-1
      image: k8s.gcr.io/pause:2.0
      resources:
        limits:
          vpc.amazonaws.com/efa: 4
```

## Additional context/links
1. K8s device plugin: https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/
5. how EKS supports EFA: https://docs.aws.amazon.com/eks/latest/userguide/node-efa.html
6. how GKE supports TPU: https://cloud.google.com/tpu/docs/kubernetes-engine-setup

**Comment by d4l3k** [comment]
Chatted with @kiukchung, Travis about this and came to the decision that we want to support this by adding a new `devices` field to Resources with a list of kubernetes resources. The current `DeviceMount` interface is a bit too low level to handle cleanly across the different schedulers.

We're using the kubernetes resource IDs since they're fairly unique and easy to standardize on.

This would look like:

```
r = Resources(
  ...,
  devices={
    "vpc.amazonaws.com/efa": 1,
    "cloud-tpus.google.com/v2": 8,
    "aws.amazon.com/neuron": 1,
  },
)
```

To support this across different schedulers we should make a new `devices.py` file with a mapping of all these highlevel device names to the underlying device mounts for schedulers that don't directly accept them.

```py
DEVICES: Mapping[str, Iterable[DeviceMount]] = {
  "aws.amazon.com/neuron": (
    DeviceMount(...),
    DeviceMount(...),
    DeviceMount(...),
  ),
}
```

down the line we can extend support for this by using entrypoints to allow registering custom devices. Keeping it centralized allows for easier adding of new devices.

Each scheduler will have to add support for the devices.

**Comment by d4l3k** [comment]
Closed by  #447
